src/engine/renderer/tr_shader.cpp

```diff
@@ -5486,22 +5486,31 @@ static shader_t *MakeShaderPermanent()
 
 	tr.numShaders++;
 
-	for ( i = 0; i < newShader->numStages; i++ )
-	{
-		if ( !stages[ i ].active )
-		{
-			break;
-		}
+	newShader->stages = nullptr;
+	newShader->lastStage = nullptr;
+
+	DAEMON_ASSERT( numStages < MAX_SHADER_STAGES );
+	DAEMON_ASSERT( MAX_SHADER_STAGES < std::numeric_limits<uint8_t>::max() );
+	DAEMON_ASSERT( MAX_TEXTURE_BUNDLES < std::numeric_limits<uint8_t>::max() );
 
-		newShader->stages[ i ] = (shaderStage_t*) ri.Hunk_Alloc( sizeof( stages[ i ] ), ha_pref::h_low );
-		*newShader->stages[ i ] = stages[ i ];
+	if ( numStages )
```

Comment:
No need for an if. The functioning should be consistent zero or nonzero

Reply:
This `if` is the solution for your other remark:

> Testing whether the begin pointer is null doesn't look good because it really depends on the details of how the array was allocated (e.g. `malloc(0)` may or may not be null)

For which I wrote:

> If there is no stage at all the code explicitly sets this pointer to `nullptr` and does no allocation at all.

Reply:
The `if` is now removed.